packages/flutter/lib/src/material/bottom_navigation_bar_theme.dart

```diff
@@ -0,0 +1,261 @@
+        && other.unselectedLabelStyle == unselectedLabelStyle
+        && other.showSelectedLabels == showSelectedLabels
+        && other.showUnselectedLabels == showUnselectedLabels
+        && other.type == type;
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<Color>('backgroundColor', backgroundColor, defaultValue: null));
+    properties.add(DiagnosticsProperty<double>('elevation', elevation, defaultValue: null));
+    properties.add(DiagnosticsProperty<IconThemeData>('selectedIconTheme', selectedIconTheme, defaultValue: null));
+    properties.add(DiagnosticsProperty<IconThemeData>('unselectedIconTheme', unselectedIconTheme, defaultValue: null));
+    properties.add(DiagnosticsProperty<Color>('selectedItemColor', selectedItemColor, defaultValue: null));
+    properties.add(DiagnosticsProperty<Color>('unselectedItemColor', unselectedItemColor, defaultValue: null));
+    properties.add(DiagnosticsProperty<TextStyle>('selectedLabelStyle', selectedLabelStyle, defaultValue: null));
+    properties.add(DiagnosticsProperty<TextStyle>('unselectedLabelStyle', unselectedLabelStyle, defaultValue: null));
+    properties.add(DiagnosticsProperty<bool>('showSelectedLabels', showSelectedLabels, defaultValue: null));
+    properties.add(DiagnosticsProperty<bool>('showUnselectedLabels', showUnselectedLabels, defaultValue: null));
+    properties.add(DiagnosticsProperty<BottomNavigationBarType>('type', type, defaultValue: null));
+  }
+}
+
+/// Applies a bottom navigation bar theme to descendant [BottomNavigationBar]
+/// widgets.
+///
+/// Descendant widgets obtain the current theme's [BottomNavigationBarTheme]
+/// object using [BottomNavigationBarTheme.of]. When a widget uses
+/// [BottomNavigationBarTheme.of], it is automatically rebuilt if the theme
+/// later changes.
+///
+/// A bottom navigation theme can be specified as part of the overall Material
+/// theme using [ThemeData.bottomNavigationBarTheme].
+///
+/// See also:
+///
+///  * [BottomNavigationBarThemeData], which describes the actual configuration
+///    of a bottom navigation bar theme.
+class BottomNavigationBarTheme extends InheritedWidget {
+  /// Constructs a bottom navigation bar theme that configures all descendant
+  /// [BottomNavigationBar] widgets.
+  ///
+  /// The [data] must not be null.
+  const BottomNavigationBarTheme({
+    Key key,
+    @required this.data,
+    Widget child,
+  }) : assert(data != null), super(key: key, child: child);
+
+  /// The properties used for all descendant [BottomNavigationBar] widgets.
+  final BottomNavigationBarThemeData data;
+
+  /// Returns the configuration [data] from the closest
+  /// [BottomNavigationBarTheme] ancestor. If there is no ancestor, it returns
+  /// [ThemeData.bottomNavigationBarTheme]. Applications can assume that the
+  /// returned value will not be null.
```

Comment:
Can they? Shouldn't we allow ThemeData.bottomNavigationBarTheme to be null (by default)?

Reply:
The pattern we have used thus far is to have the `fooTheme` properties on `ThemeData` default to non-null `FooThemeData` objects, with all `null` properties. 

https://github.com/flutter/flutter/blob/bec63a54bfad422c2a175a1d73052e67cc9785a9/packages/flutter/lib/src/material/theme_data.dart#L356

Reply:
OK